### ISSUE TITLE
[question] Validator Speed

### ISSUE THREAD

**Issue by siemvaessen** [codeforIATI/validator#34]
As we are considering to use another Validator to integrate / interface with for IATI.cloud I'd be interested to learn what the speed of validation is for this specific validator. Can anyone provide any time on say validating the full brunt of all XML files listed on IATI.cloud (as synced with Registry).

**Comment by stevieflow** [comment]
@siemvaessen by validation, do you mean schema validation?

and/or does this including validating codelist values, and/or even rulesets?

**Comment by siemvaessen** [comment]
@stevieflow schema validation as a bare minimum. Others optional. 

As part of IATI.cloud data processing it now has the ability to check if anything in an activity has changed and will only process activities that actually have changed (currently being revised as bulk of orgs publishing in the IATI format abuse the last-date-update field, we can blame SAP however).  

So next to validation (incl. schema validation including validating codelist values, and/or even rulesets) we also are interested to see if we enhance validation on activity level rather than just dataset level. 

The current IATI Validator actual as hosted by IATI validates on dataset level only. 

In short we are looking for a validator that does schema validation and later stage validating codelist values, and/or even rulesets plus the ability to validate on activity level. The option is the review the IATI Validator actual and enhance it with activity level validation. But also interested what the validator in this repo is planning to do or not planning to do.

**Comment by stevieflow** [comment]
Thanks @siemvaessen 

I know the tension between the "data file" and the "activity" is a longstanding one in IATI land.  

Given that most organisations publish many activities, the function of the standard to package together multiple activity documents into one file has proven useful to the _humans_ --> the alternative is that *everyone* publishes an XML document for each activity, is certainly feasible (and I assume no problem to the _machines_), but not in the design of IATI.  I assume the Registry would handle this without problem (as it's a data catalogue software), but the IATI guidance clearly advices against it

It's interesting that services such as d-portal serve up each Activity as a single XML document - eg: http://d-portal.org/q.xml?aid=US-EIN-300108263-GGProj:46508 - here, we can see that the attributes between ``iati-activities`` and ``iati-activity`` are slightly rearranged, and enhanced with a namespace extension ,too.  ( cc/ @notshi )

In this example, the data *originated* in one file with many activities (``dstore:dataset="globalgiving-activity"``) but d-portal then segments them.

Does IATI.cloud not output something similar? ie: activity level XML (whilst maintaining the ``iati-activities`` root element)

> In short we are looking for a validator that does schema validation and later stage validating codelist values, and/or even rulesets

This is what IATI CoVE does.  The results can be through the web interface , or at command line / API.  

CoVE doesn't do this segmentation of an IATI data file into activity-level XML documents, but it can easily validate files such as the one output by CoVE: https://iati.cove.opendataservices.coop/data/d986d6fc-6bea-415f-918d-9873917cd45f

Hence, I think the idea of " validate on activity level" is interesting, feasible and valuable for us all, but I'm not sure what you _really_ mean ;)

**Comment by notshi** [comment]
Hey all,

We've had to break up the datasets into individual activities so that we can store them in our database.

The xmls we provide are reconstructed from internal data so not exactly the same as when they went in; eg. element order would have been fixed according to the schema if it wasn't before and we use the namespace extension so we can add context to where the single activity comes from and link it back to the Registry from SAVi.

Additionally, we planned to add a link to the validator (https://github.com/data4development/IATI-data-validator/issues/49) but that didn't happen.

**Comment by siemvaessen** [comment]
Hi all, as part of [AIDA](https://aida.tools/) we are developing a new data management tool to offer users to manage their data (IATI compliant) so we are reviewing the actual validator currently managed by IATI and [based on D4D ](https://github.com/data4development/IATI-data-validator)work. We are considering a **_real time per activity validation on the fly(~ish)_** for an activity being managed (edited) by a user. 

I am wondering if IATI CoVE is able to handle _**real time per activity validation on the fly(~ish)**_ where the following userstory applies:

**userstory**
As I user **I want to** add/edit an activity in [AIDA](https://aida.tools/)as part of the Data Management Feature **so that I** can add/edit an activity as part of my larger dataset and make sure that activity is valid (as per schema and/or etc) and therefore ensuring this added/edited activity fits nicely with the larger dataset being maintained in AIDA.

In this case the user needs instant feedback from AIDA to ensure data entry is as expected (as per schema and/or etc) without waiting for the system to start its validation. Alternative route is data gets stored, validator runs and send the user a message. 

This is complementary some of the other requirements like seeing _real time per activity on the fly(~ish)_** data changes on the AIDA platform prior to publishing (aka Preview Mode). So we have a lot to consider.

Anyway, as part of our Review we may have a look at IATI CoVE as well. To what extent are IATI CoVE and [D4D Validator ](https://github.com/data4development/IATI-data-validator/) identical? They both to schema validation, codelist validation values, and/or even rulesets right? Guess I am trying to see what the fastest route is and trying to avoid rebuilding stuff that exists..
